### ISSUE TITLE
False positive `dead_code` lint when type is only used via destructuring

### ISSUE THREAD

**Issue by daniel5151** [rust-lang/rust#132874]
### Code

```Rust
struct Foo(usize);

fn get_thing<T>() -> T { todo!() }

pub fn foo() {
    let Foo(_x) = get_thing();
    
    // no warning if the type is explicitly used when calling get_thing
    // let Foo(_x) = get_thing::<Foo>();
}
```

### Current output

```Shell
warning: struct `Foo` is never constructed
 --> src/lib.rs:1:8
  |
1 | struct Foo(usize);
  |        ^^^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Desired output

```Shell
(nothing)
```

### Rationale and extra context

For context, the specific "get_thing" function I originally ran into this was with [`zerocopy::FromBytes::read_from_bytes`]( https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html#method.read_from_bytes).

### Other cases

```Rust

```

### Rust Version

```Shell
$ rustc --version --verbose
rustc 1.82.0 (f6e511eec 2024-10-15)
binary: rustc
commit-hash: f6e511eec7342f59a25f7c0534f1dbea00d01b14
commit-date: 2024-10-15
host: x86_64-unknown-linux-gnu
release: 1.82.0
LLVM version: 19.1.1
```

**Comment by m-ysk** [comment]
@daniel5151 
I've submitted a PR #133128  to resolve this issue and the reviewer thinks that the code you made as an MRE should be warned as a dead code as you can see in the following link.

https://github.com/rust-lang/rust/pull/133128#discussion_r1847471574

I also agree to the reviewer, but at the same time I'm thinking that there's a possibility that you've actually got a bug, but you've over-generalized your code when creating the above MRE because you said that you've found this issue when using `zerocopy::FromBytes::read_from_bytes`.
So, if possible, could you give me another MRE closer to your original code by using the `zerocopy` crate?

**Comment by daniel5151** [comment]
Sure. Here's a repro using `zerocopy` which is closer to the real (and IMO, totally reasonable) code I was writing:

```rust
use zerocopy; // 0.7.35
use zerocopy_derive; // 0.7.35

use zerocopy::FromBytes;

#[derive(zerocopy_derive::FromZeroes, zerocopy_derive::FromBytes)]
#[repr(C, packed)]
struct PacketHeader {
    pub kind: u8,
    pub len: u32
}

fn main() {
    let packet = [0x1, 2, 0, 0, 0, b'h', b'i'];
    let PacketHeader { kind, len } = FromBytes::read_from_prefix(&packet).unwrap();
    dbg!((kind, len));
}
```

Output:

```
   Compiling playground v0.0.1 (/playground)
warning: struct `PacketHeader` is never constructed
 --> src/main.rs:8:8
  |
8 | struct PacketHeader {
  |        ^^^^^^^^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `playground` (bin "playground") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.48s
     Running `target/debug/playground`
[src/main.rs:16:5] (kind, len) = (
    1,
    2,
)
```

**Comment by m-ysk** [comment]
Thanks. Your new repro seems reasonable to me. I'll think about how to go ahead with my PR #133128 again.

**Comment by mu001999** [comment]
this is a duplicate of #120770 and can be fixed by #142485. cc @nnethercote ([play](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=9b80326a7cfe92ed0d4bd9f930ea0854))

**Comment by jakubadamw** [comment]
This issue was fixed in 1.89 and is covered by [this test case](https://github.com/rust-lang/rust/pull/142485/changes#diff-0872c6af56e63ea694c60242e52e8fc42231e2d0b76e88fc4e09c8b5818ff543R28). It can be closed.
